connectors/dynamodb/stream/helper.go

```diff
@@ -0,0 +1,350 @@
+package stream
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodbstreams"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodbstreams/types"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/time/rate"
+)
+
+var ErrAlreadyRunning = fmt.Errorf("already running")
+var ErrNotRunning = fmt.Errorf("not running")
+
+var DefaultDescribeLimiter = rate.NewLimiter(rate.Limit(5), 5)
+
+type Limiter interface {
+	Wait(context.Context) error
+}
+
+type Logger interface {
+	Debug(string, ...any)
+	Info(string, ...any)
+	Warn(string, ...any)
+	Error(string, ...any)
+}
+
+type shardRequestResult struct {
+	shards []types.Shard
+	err    error
+}
+
+type dynamodbStreamHelper struct {
+	streamClient          *dynamodbstreams.Client
+	streamARN             string
+	shardRequestCh        chan chan shardRequestResult
+	describeStreamLimiter Limiter
+	logger                Logger
+	skipEmpty             int
+
+	running atomic.Bool
+	once    sync.Once
+	done    chan struct{}
+	err     error
+}
+
+type HelperOptions struct {
+	DescribeStreamLimiter Limiter
+	Logger                Logger
+	SkipEmpty             int // For use when searching for approximate latest
+}
+
+func NewDynamodbStreamHelper(streamClient *dynamodbstreams.Client, streamARN string, optsFn ...func(*HelperOptions)) *dynamodbStreamHelper {
+	opts := HelperOptions{
+		DescribeStreamLimiter: DefaultDescribeLimiter,
+		Logger:                slog.Default(),
+		SkipEmpty:             5,
+	}
+	for _, fn := range optsFn {
+		fn(&opts)
+	}
+	return &dynamodbStreamHelper{
+		streamClient:          streamClient,
+		streamARN:             streamARN,
+		shardRequestCh:        make(chan chan shardRequestResult), // We don't close this, but that is ok
+		describeStreamLimiter: opts.DescribeStreamLimiter,
+		logger:                opts.Logger,
+		skipEmpty:             opts.SkipEmpty,
+		done:                  make(chan struct{}),
+	}
+}
+
+func (s *dynamodbStreamHelper) Start(ctx context.Context) error {
+	err := ErrAlreadyRunning
+	s.once.Do(func() {
+		s.running.Store(true)
+		go func() {
+			s.err = s.startGetShards(ctx)
+			s.running.Store((false))
+			close(s.done)
+		}()
+		err = nil
+	})
+	return err
+}
+
+func (s *dynamodbStreamHelper) Wait() error {
+	<-s.done
+	return s.err
+}
+
+func (s *dynamodbStreamHelper) GetShards(ctx context.Context) ([]types.Shard, error) {
```

Comment:
Nitpick: There's a function with the exact same name in the same file but lower-case first letter, yet the way they do things is very different. Can the function name be more descriptive?

Reply:
this is the public one, under the hood it ends up using the other one